heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,249 @@
+            + "existing containers, but the scheduler does not support scaling, aborting. "
+            + "Existing packing plan: %s, proposed packing plan: %s",
+        newContainerCount, removableContainerCount, existingPackingPlan, proposedPackingPlan);
+    assertTrue(newContainerCount + removableContainerCount == 0 || scalableScheduler.isPresent(),
+        message);
+
+    SchedulerStateManagerAdaptor stateManager = Runtime.schedulerStateManagerAdaptor(runtime);
+
+    // assert found PackingPlan is same as existing PackingPlan.
+    validateCurrentPackingPlan(existingProtoPackingPlan, topologyName, stateManager);
+
+    // fetch the topology, which will need to be updated
+    TopologyAPI.Topology updatedTopology =
+        getUpdatedTopology(topologyName, proposedPackingPlan, stateManager);
+
+    // request new resources if necessary. Once containers are allocated we should make the changes
+    // to state manager quickly, otherwise the scheduler might penalize for thrashing on start-up
+    if (newContainerCount > 0) {
+      scalableScheduler.get().addContainers(containerDelta.getContainersToAdd());
+    }
+
+    // update parallelism in updatedTopology since TMaster checks that
+    // Sum(parallelism) == Sum(instances)
+    logFine("Deleted existing Topology: %s", stateManager.deleteTopology(topologyName));
+    logFine("Set new Topology: %s", stateManager.setTopology(updatedTopology, topologyName));
+
+    // update packing plan to trigger the scaling event
+    logFine("Deleted existing packing plan: %s", stateManager.deletePackingPlan(topologyName));
+    logFine("Set new PackingPlan: %s",
+        stateManager.setPackingPlan(proposedProtoPackingPlan, topologyName));
+
+    // delete the physical plan so TMaster doesn't try to re-establish it on start-up.
+    logFine("Deleted Physical Plan: %s", stateManager.deletePhysicalPlan(topologyName));
+
+    if (removableContainerCount > 0) {
+      scalableScheduler.get().removeContainers(containerDelta.getContainersToRemove());
+    }
+  }
+
+  @VisibleForTesting
+  TopologyAPI.Topology getUpdatedTopology(String topologyName,
+                                          PackingPlan proposedPackingPlan,
+                                          SchedulerStateManagerAdaptor stateManager) {
+    TopologyAPI.Topology updatedTopology = stateManager.getTopology(topologyName);
+    Map<String, Integer> proposedComponentCounts = proposedPackingPlan.getComponentCounts();
+    return mergeTopology(updatedTopology, proposedComponentCounts);
+  }
+
+  @VisibleForTesting
+  void validateCurrentPackingPlan(PackingPlans.PackingPlan existingProtoPackingPlan,
+                                  String topologyName,
+                                  SchedulerStateManagerAdaptor stateManager) {
+    PackingPlans.PackingPlan foundPackingPlan = stateManager.getPackingPlan(topologyName);
+    assertTrue(foundPackingPlan.equals(existingProtoPackingPlan),
```

Comment:
By multiple agents, do you mean concurrent update requests? 
In my opinion this check should be removed. Jut in case scheduler plan is not in sync with persisted plan, scheduler would be able to persist the new desired plan. This check enforces topology restart.